app/services/sync_salesforce_program.rb

```diff
@@ -1,87 +1,313 @@
+  # properly created. They are not for issues with Section enrollment for an individual
+  # Participant
+  class SectionSetupError < StandardError; end
+  class CreateSectionError < SectionSetupError; end
+  class MissingSectionError < SectionSetupError; end
 
-  def initialize(salesforce_program_id:, force_zoom_update: false)
-    @sf_program_id = salesforce_program_id
-    @sf_program = nil
+  FailedParticipantInfo = Struct.new(:participant_id, :email, :first_name, :last_name, :error_detail, keyword_init: true)
+
+  def initialize(salesforce_program, max_duration_seconds, force_zoom_update=false)
+    @salesforce_program = salesforce_program
+    @max_duration_seconds = max_duration_seconds
     @force_zoom_update = force_zoom_update
     @failed_participants = []
   end
 
+  # Handles all syncing logic for all Participants in a Program, such as:
+  # 1. creating them when they become Enrolled
+  # 2. syncing their email and other Contact info
+  # 3. moving them between sections if their enrollment changes
+  # 4. unenrolling them if they become Dropped
+  # 5. adding them to the proper TA Caseload section
+  # 6. syncing their Zoom links
+  #
+  # Note: Discord is separate b/c it's a bot.
   def run
-    Honeycomb.add_field_to_trace('salesforce.program.id', @sf_program_id)
-    Honeycomb.add_field('sync_salesforce_program.force_zoom_update', @force_zoom_update)
-
-    # If you try running a sync in an environment that doesn't have the corresponding local course(s),
-    # this is a NOOP. This is mostly in place to prevent developers from accidentally sync'ing another
-    # dev's test program which would invalidate the signup tokens and other stuff.
-    if courses_for_program.blank?
-      Honeycomb.add_field('alert.sync_salesforce_program.missing_courses', sf_program.inspect)
-      raise SyncSalesforceProgramError, "Missing Course models for Salesforce Program: #{sf_program.inspect}"
+    @start_time = Time.now.utc
+    Honeycomb.start_span(name: 'sync_salesforce_program.run') do
+      @salesforce_program.add_to_honeycomb_trace()
+      Honeycomb.add_field('sync_salesforce_program.force_zoom_update', @force_zoom_update)
+      Rails.logger.debug("Started sync for Program Id: #{@salesforce_program.sfid}")
+
+      if @salesforce_program.accelerator_course.blank? || @salesforce_program.lc_playbook_course.blank?
+        msg = "Missing local Course models for Salesforce Program ID: #{@salesforce_program.sfid}"
+        Honeycomb.add_field('alert.sync_salesforce_program.missing_courses', true)
+        Honeycomb.add_field('sync_salesforce_program.participant.skip_reason', msg)
+        raise MissingCourseModelsError, msg
+      end
+
+      @participants = ParticipantSyncInfo.diffs_for_program(@salesforce_program)
+      @count = @participants.count
+      Honeycomb.add_field('sync_salesforce_program.participants.count', @count)
+
+      if @force_zoom_update == false && @participants.find { |p| p.changed? }.nil?
```

Comment:
i think this is clearer what's happening in the second check
```suggestion
      if @force_zoom_update == false && @participants.any? { |p| p.changed? }
```